http3/client.go

```diff
@@ -201,6 +262,34 @@ func (c *client) RoundTrip(req *http.Request) (*http.Response, error) {
 	return rsp, rerr.err
 }
 
+func (c *client) requestCancelled(str quic.Stream, req *http.Request, reqDone <-chan struct{}) {
+	select {
+	case id := <-c.goawayChan:
```

Comment:
This won't work if there's more than a single request.

Reply:
Do you mean multiple GOAWAY frames or multiple client requests on the same connection? I have found a way to handle multiple GOAWAY frames, but this code should work for multiple requests (further down in that code it puts the id back on to the channel so that other requests can read it as well)

Reply:
Please don't do that, that's highly non-obvious. It would be a lot cleaner to have a `chan struct{}`, close that channel, and then read from a member variable of the client.

Reply:
Yes, that is true. I have done as you said, with a slight difference: since the client can receive multiple go away frames, I am using a slice of `chan struct{}` (not one for each client! just one for each goaway frame sent). I cannot see any alternative that does not involve managing client channels with a map of sorts. Is this ok?

Reply:
I don't think that's necessary. I'd assume something along these lines makes sense. You add member variables to the client:
```go
type client {
    // other stuff
    goawayChan chan struct{}
    goawayStreamNum protocol.StreamID
}
```

When receiving a GOAWAY frame, you

- `if c.goawayStreamNum == 0`, this is the first GOAWAY frame, so you `close(c.goawayChan)`
- `else`, you check that the new GOAWAY frame carries a smaller stream number than the one previously received (and throw an error otherwise), and update the value of `c.goawayStreamNum`

Does that make sense?

Reply:
Yes, and that is a solution that I considered. However, while it works for when
- Client is about to reach `requestCancelled`
- Client `requestCancelled` goroutine is already sleeping when the first goaway frame is received
it will not work when client the `requestCancelled` goroutine is sleeping and the first goaway frame has already been received, and the streamID was < than that goaway frame. In that case, it will still be sleeping, and since we closed the channel there is no way to communicate with them if another goaway frame is received. If you believe handling this case is not necessary, I can go with the simpler one channel solution (closing client connections on goaway is optional after all)
I have a demo of what I am doing in the code here: https://play.golang.org/p/boPrANTKCnM

Reply:
The `[]chan struct` looks kind of complicated, I'd prefer to avoid that.

What about using a `sync.Cond` and then using a `Broadcast()` every time a GOAWAY is received?

Reply:
That was also something I tried, however `sync.Cond` uses `Wait()`, which I couldn't see a way of using alongside the existing channel mechanism we use (`<-reqDone`). sync.Cond is kind of inflexible in this aspect. I couldn't see a way of feeding it a context somehow; if that were possible I think I could make it work. Looking through docs right now. Basically we need a mutex that takes a context: if we have that then `sync.Cond` is workable.

Reply:
Ok, I have only one possible solution through `sync.Cond`, though it is arguably more obtuse than my own. Essentially we use two goroutines, the original `requestCancelled` goroutine along with another one that uses `sync.Cond.Wait()`, and when the `requestCancelled` goroutine finishes, we `Broadcast()` to it a special termination signal (say -2) that tells that goroutine to terminate. I am not sure of the safety of `Broadcast()` concurrently, but it should be workable. 

Spinning two goroutines is mandatory for `sync.Cond`, as it uses `Wait()`  semantics instead of channels, so you need it on its own goroutine.

The only other option besides this and my channel slice is a map of client goroutines (at least after my analysis). I just can't see a way of telling the `sync.Cond` to stop other than to Broadcast to it every time

Reply:
@lucas-clemente Which of these options do you prefer?

Reply:
I think this can be solved a lot simpler. So if I see this correctly, the problem we're trying to solve is how to deal with requests that are blocked on `OpenStreamSync` when a GOAWAY is received. Requests that are already sent to the server don't need to be canceled, and requests for which the server is already sending a response can also just continue as if nothing happened.

I think it would be a valid behavior to return an error from all those requests, even if they might have gotten a stream ID lower than the limit in the GOAWAY frame. GOAWAY is used for graceful shutdown, after all.

Reply:
The issue is when multiple GOAWAYs are sent. When just one is sent, it is no issue. Multiple GOAWAYs require something more complex to handle them. We can always just only parse the first one, and have the server reject any further client conns itself while also sending the GOAWAY. This would not be against the spec because it says that 

> Some requests might already be in transit. If the client has already sent requests on streams with a Stream ID greater than or equal to that indicated in the GOAWAY frame, those requests will not be processed and MAY be retried by the client on a different connection. The client **MAY** cancel these requests. *It is RECOMMENDED that the server explicitly reject such requests (see Section 4.1.2) in order to clean up transport state for the affected streams.*

Emphases mine